Add TriCore Architecture

b1llow opened the pull request:
Changes based on https://github.com/TriDis/capstone/tree/tricore, adaptations for new capstone versions and support for current versions of the TriCore instruction set.

The current `tc1.6.2` version of instruction set disassembly is almost complete. Other versions actually overlap with `tc1.6.2` for the most part, with only a small amount of extra content, which is still under construction.

Then it should be mentioned that basically all the `.inc` is generated using the modified `llvm-tblgen` from https://github.com/Rot127/llvm-capstone/tree/auto-sync-tblgen, except for `TriCoreGenDisassemblerTables.inc`, which modified a small part of `DecodeInstruction` and `DecodeToMCInst`

TODO:

- [x] tc1.1
- [x] tc1.2
- [x] tc1.3
- [x] tc1.3.1
- [x] tc1.6
- [x] tc1.6.1
- [x] tc1.6.2
- [x] tc1.1 tests
- [x] tc1.6.2 basic tests
- [x] others tests

sidneyp committed:
Add TriCore Architecture

sidneyp committed:
Modify Makefiles for TriCore architecture

sidneyp committed:
Fix Disassembler and TableGen Files of TriCore

sidneyp committed:
Map instruction and fix part of InstPrinter

sidneyp committed:
Improve TriCore InstPrinter and Mapping

sidneyp committed:
Fix TriCore GenAsmWriter and InstPrinter

sidneyp committed:
Add group name maps

sidneyp committed:
Fix error of undeclared instructions and registers

sidneyp committed:
Fix all errors in TriCoreDisassembler

sidneyp committed:
Fix all errors in TriCoreInstPrinter

sidneyp committed:
Fix all compiler errors

sidneyp committed:
Remove all compiler warnings

sidneyp committed:
Fix printing MemSrc

sidneyp committed:
Adjust printing of Registers and upgrade TriCore test

sidneyp committed:
Transfer modifications of TriDis/llvm-tricore on Sep 20, 2016

sidneyp committed:
Transfer modifications of TriDis/llvm-tricore on Oct 05, 2016

sidneyp committed:
Fix memory access printing and clean unused functions

sidneyp committed:
Adjustments in TriCore and add more instructions into tests/test_tricore.c

sidneyp committed:
Fix TriCore Mapping

sidneyp committed:
Fix TriCore register mapping

sidneyp committed:
Add document with available TriCore instructions.

sidneyp committed:
Transfer modifications of TriDis/llvm-tricore on Feb, 04 2017

sidneyp committed:
Update document with newly available TriCore instructions

b1llow committed:
Fix rebase error

b1llow committed:
Fix build and test

b1llow committed:
fix arch_configs[CS_ARCH_TRICORE]

aquynh commented:
Very nice work! From which llvm is this, please?

b1llow commented:
> Very nice work! From which llvm is this, please?

Maybe https://github.com/TriDis/llvm-tricore

aquynh commented:
Is that unmaintained repo?



On Sat, Mar 11, 2023, 05:49 billow ***@***.***> wrote:

> Very nice work! From which llvm is this, please?
>
> Maybe https://github.com/TriDis/llvm-tricore
>
>
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/capstone-engine/capstone/pull/1973#issuecomment-1464591176>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNQNYBHORMKVOVWAY6C4KDW3OVXJANCNFSM6AAAAAAVXABL74>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

b1llow commented:
Yes, and I meant to use this as a starting point only. We could probably maintain TriCore's TableGen and TriCoreDisassembler code ourselves.

aquynh commented:
Can we assume that this arch is no longer developed? (Unlike Arm, x86, etc)

On Sat, Mar 11, 2023, 06:19 billow ***@***.***> wrote:

> Yes, and I meant to use this as a starting point only. We could probably
> maintain TriCore's TableGen and TriCoreDisassembler code ourselves.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/capstone-engine/capstone/pull/1973#issuecomment-1464630233>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNQNYGVZ3B2YHQC2EC5N3DW3OZJRANCNFSM6AAAAAAVXABL74>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

b1llow commented:
I think it's just not as widely used.

b1llow commented:
ping @XVilka

XVilka commented:
@imbillow @imbillow this arch is indeed not widely used but *quite common* in the automotive and few other industries. This architecture is still being improved with ISA changes slightly from version to version, latest one being 1.6.2:
- https://wiki.qemu.org/Documentation/Platforms/TriCore

The aforementioned LLVM fork isn't complete and abandoned, and Infineon didn't indicate that they plan to implement Tricore support in the mainline LLVM or even some open fork. Thus, I suggest we focus only on the capstone and maintain it separately from the LLVM things.

XVilka commented:
@imbillow note, there are slight differences for some instructions between different ISA versions, we should add the way to specify the particular ISA in the API.

XVilka commented:
@Rot127 could you please also take a look, in case it will interfere with the auto-sync work?

Rot127 commented:
@XVilka Looks good to me.
Only the architectures which are generated with `auto-sync` will be adjusted to each other. Since you mentioned that Tricore will not be in the LLVM mainline (and is therefore not a candidate for `auto-sync`) there's no need to adjust it to the `auto-sync` design.

b1llow committed:
add TriCore tablegen files

b1llow committed:
add more `ADD`

b1llow committed:
add `ABS`

b1llow committed:
Add more multiclass

b1llow committed:
add more `ADD`

b1llow committed:
add `AND`

b1llow committed:
add `cache` inst

b1llow committed:
add `cadd`

b1llow committed:
add:
- `CALL*`
- `CL*`
- `CMOV*`
- `CMPSWAP_W`
- `CRC*`
- `CSUB*`
- `DEBUG*`

b1llow committed:
add:
- `DEXTR*`
- `DISABLE*`
- `DSYNC*`
- `DVADJ*`
- `DIV*`
- `DVINIT*`
- `DVSTEP*`
- `ENABLE*`
- `EQ*`

b1llow committed:
Changes to be committed:
	modified:   arch/TriCore/TriCoreInstrFormats.td
	modified:   arch/TriCore/TriCoreInstrInfo.td

b1llow committed:
modified:   arch/TriCore/TriCoreInstrInfo.td

b1llow committed:
`ld*`

b1llow committed:
- `LD*`
- `LEA*`
- `LHA*`
- `LOOP*`
- `LT*`

b1llow committed:
`MADD*`

b1llow committed:
fix `MADD*`

b1llow committed:
`MAX*` `MFCR` `MIN*`

b1llow committed:
`MOV*`

b1llow committed:
`MSUB*`

b1llow committed:
`MUL*` `NAND*` `NE*` `NOP` `NOR` `OR`

b1llow committed:
modified:   arch/TriCore/TriCoreInstrInfo.td

b1llow committed:
RSUB* SAT* SEL* SH*

b1llow committed:
SHA* ST*

b1llow committed:
ST SUB SYSCALL XOR XNOR

b1llow committed:
update `TriCore*.inc`

b1llow committed:
add FPU Instructions

b1llow committed:
fix build

b1llow committed:
`Decode*Instruction`

b1llow committed:
fix `DecodeInstruction`

b1llow committed:
add `tricore` to `cstool`

b1llow committed:
add `tricore` to `cstest`

b1llow committed:
fix

b1llow committed:
fix

kabeor commented on cs.c:
```diff
@@ -368,9 +381,9 @@ bool CAPSTONE_API cs_support(int query)
-				    (1 << CS_ARCH_WASM)  | (1 << CS_ARCH_BPF)        |
-				    (1 << CS_ARCH_SH));
+				    (1 << CS_ARCH_RISCV) | (1 << CS_ARCH_MOS65XX)    |
+				    (1 << CS_ARCH_WASM)  | (1 << CS_ARCH_BPF))       |
```
plz move the last `)`to next line so we can avoid gcc warning.

kabeor commented:
plz check CIFuzz.

```
Test MC/RISCV/insn-riscv64.s.cs
Test MC/TriCore/ADC_Background_Scan_1_KIT_TC275_LK.s.cs
fail CS_ARCH_TRICORE CS_MODE_TRICORE
Traceback (most recent call last):
  File "./test_corpus.py", line 134, in <module>
    test_file(fname.strip())
  File "./test_corpus.py", line 125, in test_file
    fout.write(unichr(mc_modes[(arch, mode)]))
KeyError: ('CS_ARCH_TRICORE', 'CS_MODE_TRICORE')
2023-03-25 03:08:35,059 - root - ERROR - Building fuzzers failed.
2023-03-25 03:08:35,060 - root - ERROR - Error building fuzzers for (commit: b01e5c9b2261bfad775d21b7c4b62fb5a59e229f, pr_ref: refs/pull/1973/merge).
```

Rot127 commented:
Could you please check my messages in Rizins Mattermost? Depending what you say regarding the `auto-sync` compatibility, I'd like to review this as well.

b1llow committed:
- add `tricore` to python binding
- try fix `test_corpus.py`

b1llow committed:
add more `tricore` cstest

b1llow committed:
add `sign_ext` `zero_ext`

b1llow committed:
fix

b1llow committed:
fix `DecodeRegisterClass`
- `DecodeABSInstruction`
- `DecodeRLCInstruction`
- `DecodeSSRInstruction`
- `DecodeSRCInstruction`

b1llow committed:
fix `DecodeRegisterClass`

b1llow committed:
fix `BO` `BOL`

b1llow committed:
fix `SRRS`

b1llow committed:
fix

b1llow committed:
fix `RRPW`

b1llow committed:
fix `INSERT_rrpw`

b1llow committed:
fix `RR`

b1llow committed:
fix

b1llow committed:
fix `mov.a` `mov.d`

b1llow committed:
fix `imask`

b1llow committed:
Whether to call `a10` a `sp`

b1llow committed:
add `MTCR`

b1llow committed:
add `BO`

XVilka commented on cstool/cstool.c:
```diff
@@ -105,6 +105,7 @@ static struct {
 	{ "sh4abe", CS_ARCH_SH, CS_MODE_BIG_ENDIAN | CS_MODE_SH4A | CS_MODE_SHFPU },
 	{ "sh4al-dsp", CS_ARCH_SH, CS_MODE_LITTLE_ENDIAN | CS_MODE_SH4A | CS_MODE_SHDSP | CS_MODE_SHFPU },
 	{ "sh4al-dspbe", CS_ARCH_SH, CS_MODE_BIG_ENDIAN | CS_MODE_SH4A | CS_MODE_SHDSP | CS_MODE_SHFPU},
+	{"tricore", CS_ARCH_TRICORE, CS_MODE_TRICORE},
```
Missing space

XVilka commented on docs/TricoreInstructions.ods:
I think this file could be removed

b1llow committed:
docs: Refactor project structure and remove unnecessary file.

b1llow commented:
There should be special handling for different versions of ISA that hasn't been done yet. Currently, only TC1.6.2 is being referenced. Also, the testing may not be comprehensive enough, and I feel that there are still quite a few bugs present.

b1llow committed:
fix: TriCore architecture disassembly codes

- Rename `ISLR_post_increment` to `ISLR_pos` for clarity
- Fix register decoding for TriCore architecture in `TriCoreDisassembler.c`
- Add new file `LoadStore.s.cs` to `suite/MC/TriCore`

b1llow committed:
fix: TriCore instruction decoding and printing.

- Modify TriCore instructions to use bracket syntax and offsets for better clarity
- Add off4_fixup and printSExtImm_n

b1llow committed:
fix: decode error

b1llow committed:
Fix: TriCore instruction operations and decoding.

b1llow committed:
fix: TriCore Instruction Formats and Printing

- Add new operand type `off18imm`
- Add `printOff18Imm` function for printing specific immediate value

b1llow committed:
refactor: Refactor TriCore instructions and operands

- Add new immediate operands and refactor code for better readability in TriCoreInstrInfo
- Update InstPrinter functions to handle new immediate operands and remove unused function

b1llow committed:
feat: Fix bugs and update instructions for TriCore architecture.

- Fix bug in `printDisp8Imm` function in `TriCoreInstPrinter.c`
- Add new `CALL_sb` instruction to `TriCoreInstrInfo.td`
- Reorder instruction definitions and operands in `TriCoreInstrInfo.td`

b1llow committed:
fix: decode `j` `call` `loop`

b1llow committed:
unique tests

b1llow committed:
fix: `disp` print and fill

b1llow committed:
just add TriCoreISA enum

b1llow committed:
add tricore Predicates

b1llow committed:
cleanup

b1llow committed:
add some tricore v1.1 inst

b1llow committed:
fix `cachea`

b1llow committed:
refactor: Refactor TriCore register names.

- Rename TriCore register names for better readability
- Update TriCore instruction information

b1llow committed:
refactor: Improve TriCore instruction handling.

- Add new instructions for TriCore architecture
- Implement changes to TriCoreInstrInfo.td

b1llow committed:
refactor: Add support for new TriCore instructions and constraints.

- Add support for V110 and V120 in various LD instructions
- Define new multiclass for code reuse
- Restructure LD_A_v120 in favor of LD_A with Requires constraint
- Add new defs for LT_U with suffixes for V110 support

b1llow committed:
refactor: Update TriCore instruction requirements

- Update `Requires` directives for several instructions to reflect changes in hardware versions

b1llow committed:
refactor: Refactor TriCore instruction decoding and register definition.

- Update TriCore processor register definitions with auto-generated file `TriCoreGenCSRegEnum.inc`
- Add several new TriCore processor instructions with auto-generated file `TriCoreGenCSInsnEnum.inc`
- Update TriCore_OP_GROUP enumeration with auto-generated file `TriCoreGenCSOpGroup.inc`
- Rename and restructure TriCore processor register classes
- Remove unused register class definitions and related code

b1llow committed:
add `tricore_feature` support

b1llow committed:
feat: Add and remove TriCore instructions.

- Add 3 new TriCore instructions
- Remove TriCore instruction "TriCore_INS_INIT"
- Alphabetized and rearranged various TriCore instructions
- Commented out code remains in the diff but is not part of the program.

b1llow committed:
refactor: Add new TriCore instructions and remove deprecated ones.

- Remove deprecated instructions
- Update comments and formatting

XVilka commented:
@imbillow could you please also maintain a very brief TODO/status in the PR description?

b1llow commented:
> @imbillow could you please also maintain a very brief TODO/status in the PR description?

done

b1llow committed:
refactor: Refactor TriCore instruction definitions and mappings

- Add new `multiclass mI_MADDRsh_` to handle `MADDR_H` and `MADDRS_H` in `arch/TriCore/TriCoreInstrInfo.td`

b1llow committed:
refactor: Improve Architecture Instruction Information.

- Refactor TriCore instruction information file
- Simplify and optimize certain instructions

b1llow committed:
refactor: Refactor TriCore instruction classes and operands

b1llow committed:
Optimize TriCore instruction information.

- Refactor TriCore instructions
- Simplify operand encoding for better readability

b1llow committed:
Optimize TriCore instruction information.

- Refactor and optimize instruction encoding for TriCore architecture
- Improve performance and reduce code size by simplifying opcodes and encoding logic
- Update instruction definitions in TriCoreInstrInfo.td file
- Omit some large but non-essential changes in diff summary for readability

b1llow committed:
refactor: Improve TriCore instruction definitions in architecture file

- Refactor architecture specific code for TriCore
- Update TriCore instruction information in [arch/TriCore/TriCoreInstrInfo.td]
- Improve performance and readability of relevant functions

b1llow committed:
refactor: Improve instruction handling for TriCore architecture

b1llow committed:
Add support for TriCore V162 and new instructions/operands.

- Add new instruction `MOVZ_A`, remove instruction `NOT`, and add several new multiply and multiply-subtract instructions
- Move `multiclass mISR_1` and `multiclass mISYS_0` to separate file and fix typo in `rfe` instruction in `mISYS_0`
- Add support for new CPU feature `TriCore_FEATURE_HasV162` and update relevant inc files.

b1llow committed:
Refactor: Optimize TriCore instruction information.

- Refactor TriCore instruction info
- Improve code organization
- Optimize instruction handling
- Eliminate unnecessary code duplication

b1llow committed:
refactor: Improve TriCore floating-point operations in instruction set

b1llow committed:
refactor: Optimize TriCore instructions in arch/TriCore/TriCoreInstrInfo.td

- Add missing instructions and update existing ones

b1llow committed:
fix

b1llow committed:
feat: Add support for TriCore feature bits and new architectures

- Add support for new Tricore architectures
- Clean up redundant instructions definitions
- Modify architecture options for the TRICORE mode
- Update disallowed modes for Tricore architecture

b1llow committed:
feat: Refactor and improve triCore platform support

b1llow committed:
feat: Update TriCore processor support and architecture modes

- Add support for TRICORE architecture modes 110, 120, 130, 131, 160, 161, and 162 in test_corpus.py
- Change the TriCore mode from `CS_MODE_TRICORE` to `CS_MODE_TRICORE_162` in `Tricore/*.s.cs`

b1llow committed:
fix

b1llow committed:
add some tc162 tests

b1llow committed:
fix TriCoreDisassembler.c from tests

b1llow committed:
fix tricore tests

b1llow committed:
feat: Update Tricore assembly code and disassembler logic.

- Add new Tricore test `tc110.s.cs`

b1llow committed:
add tc110 tests and fix tricore decode

b1llow committed:
fix `CADD` `CSUB`

b1llow committed:
fix tc110 test and fix decode

b1llow committed:
fix `RCR` printer

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix

b1llow committed:
fix tc1.1 tests

b1llow committed:
fix tc1.6.2 tests

b1llow committed:
fix tc1.6.2 tests

b1llow committed:
add tests

b1llow committed:
Update TriCore instructions in TriCoreInstrInfo.td

b1llow committed:
fix tests

b1llow committed:
fix tests

b1llow committed:
fix tests

b1llow committed:
fix `TRICORE_GENERIC` inst

b1llow committed:
fix tests

b1llow commented:
Now our disassembled mnemonic and op_str are pretty good, but inst_detail is not quite as good

b1llow committed:
fix `TriCore_getRegisterName`

b1llow committed:
cleanup

b1llow committed:
fix `TRICORE_OP_MEM`

XVilka commented:
@kabeor could you please take a look at this to give the early feedback, since majority of the code was written, also to allow the CI jobs to run?

b1llow committed:
fix `TriCoreGenCSMappingInsnName.inc`

b1llow committed:
Update tricore `.inc`

XVilka commented:
@imbillow see also:
```
Traceback (most recent call last):
  File "./test_corpus.py", line 149, in <module>
    test_file(fname.strip())
  File "./test_corpus.py", line 64, in test_file
    "CS_MODE_TRICORE_110": CS_MODE_TRICORE_110,
NameError: global name 'CS_MODE_TRICORE_110' is not defined
2023-04-10 14:22:54,188 - root - ERROR - Building fuzzers failed.
```

b1llow committed:
fix `test_corpus.py`

b1llow committed:
fix fuzz

b1llow committed:
fix fuzz

b1llow commented:
Because I'm not very familiar with Python 2, so I converted the `test_corpus.py` file to Python 3 syntax, resulting in the new `test_corpus3.py`. Their functionalities should be essentially the same.

However, the `cifuzz` GitHub action relies on `google/oss-fuzz` to complete the fuzz test, so it won't use `test_corpus3.py`. But I think it should still be useful, so I submitted it anyway.

XVilka commented:
@kabeor what do you think about such script conversion?

kabeor commented:
It's really great! Let me think about how to make the best use of it later.

Rot127 commented:
I hadn't time yet to add the `CAPSTONE_DIET` guards to the LLVM backends. So they are not in here as well.
@XVilka @kabeor Should this be done asap so it is part of this PR as well? Or does it have a few more days time and we maybe  update TriCore later?

XVilka commented:
@Rot127 I think we can wait, yes. @imbillow what is your opinion?

kabeor commented:
I think we can wait, and then I plan to release rc2.

b1llow commented:
Yes, I can also wait for a few days; I'm not in a big hurry.
@Rot127 @XVilka

Rot127 commented:
@imbillow `CAPSTONE_DIET` is added now. The `inc` files have everything string like removed if the flag is set.

Rot127 commented:
Note that the `AsmWrite.inc` might not generates reproducible. Shouldn't be a problem since it should be in itself consistent. But you can also just apply the guards to be sure.

b1llow commented:
@Rot127 

It looks like there are some problems with code generation without the `--printerLang=CCS` parameter.

```
llvm-tblgen: /home/aya/Source/llvm-capstone/llvm/utils/TableGen/CodeGenTarget.cpp:525: void llvm::CodeGenTarget::ComputeInstrsByEnum() const: Assertion `InstrsByEnum.size() == Insts.size() && "Missing predefined instr"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/aya/.local/bin/llvm-tblgen --gen-disassembler -I /home/aya/Source/llvm-project/llvm/include -I /home/aya/Source/capstone/arch/TriCore/ /home/aya/Source/capstone/arch/TriCore//TriCore.td
 #0 0x000055c6c00430b6 (/home/aya/.local/bin/llvm-tblgen+0x5ea0b6)
 #1 0x000055c6c004363b (/home/aya/.local/bin/llvm-tblgen+0x5ea63b)
 #2 0x000055c6c0041663 (/home/aya/.local/bin/llvm-tblgen+0x5e8663)
 #3 0x000055c6c0043e11 (/home/aya/.local/bin/llvm-tblgen+0x5eae11)
 #4 0x00007f36d8e51f50 (/usr/lib/libc.so.6+0x38f50)
 #5 0x00007f36d8ea08ec (/usr/lib/libc.so.6+0x878ec)
 #6 0x00007f36d8e51ea8 raise (/usr/lib/libc.so.6+0x38ea8)
 #7 0x00007f36d8e3b53d abort (/usr/lib/libc.so.6+0x2253d)
 #8 0x00007f36d8e3b45c (/usr/lib/libc.so.6+0x2245c)
 #9 0x00007f36d8e4a9f6 (/usr/lib/libc.so.6+0x319f6)
#10 0x000055c6bfc6a270 (/home/aya/.local/bin/llvm-tblgen+0x211270)
#11 0x000055c6bfa9b7c3 (/home/aya/.local/bin/llvm-tblgen+0x427c3)
#12 0x000055c6bfcbb3a8 (/home/aya/.local/bin/llvm-tblgen+0x2623a8)
#13 0x000055c6bfcbac7a (/home/aya/.local/bin/llvm-tblgen+0x261c7a)
#14 0x000055c6bfd12000 (/home/aya/.local/bin/llvm-tblgen+0x2b9000)
#15 0x000055c6bff135b3 (/home/aya/.local/bin/llvm-tblgen+0x4ba5b3)
#16 0x000055c6c007163e (/home/aya/.local/bin/llvm-tblgen+0x61863e)
#17 0x000055c6bff133f9 (/home/aya/.local/bin/llvm-tblgen+0x4ba3f9)
#18 0x00007f36d8e3c790 (/usr/lib/libc.so.6+0x23790)
#19 0x00007f36d8e3c84a __libc_start_main (/usr/lib/libc.so.6+0x2384a)
#20 0x000055c6bfa84125 (/home/aya/.local/bin/llvm-tblgen+0x2b125)
```

b1llow commented:
Then there are some `unterminated #ifdef`.

Rot127 commented:
Weird, I can't reproduce this with ARM.

Rot127 commented:
Let me quickly finish up a CS commit and I take a look at it.

Rot127 commented:
Just fixed the missin `#endif`. But I still can't reproduce the bug from before.
I  used the `td` files from https://github.com/imbillow/capstone/tree/tricore
Did you made changes to your `td` files which are not yet pushed?

b1llow committed:
update TriCoreGenAsmWriter.inc

b1llow commented:
Sorry about that. It's actually because I ran `llvm-tblgen` with this parameter `-I llvm-project/llvm/include`, which caused the program to `Assertion failed`. But when I use the correct parameter `-I llvm-capstone/llvm/include`, the program behaves normally.

XVilka commented:
@imbillow so I assume this PR is finished, right? @kabeor could you please take a look again?

b1llow commented:
> @imbillow so I assume this PR is finished, right? @kabeor could you please take a look again?

Yes. And since the `.inc` files and `.s.cs` files are almost entirely generated, they can largely be ignored when reviewing.

kabeor commented:
Great，LGTM. By the way, could you please add TriCore in *Support multiple hardware architectures* in README? @imbillow 

I think it's ready to merge, any suggestion? @aquynh

b1llow committed:
add TriCore in README

Rot127 commented:
> Yes. And since the .inc files and .s.cs files are almost entirely generated, they can largely be ignored when reviewing.

Which `inc` files were not generated?

Rot127 commented:
Also I suggest to add the TriCore `td` files to Capstones LLVM fork.

b1llow commented:
@Rot127
> Then it should be mentioned that basically all the `.inc` is generated using the modified `llvm-tblgen` from [Rot127/llvm-capstone@`auto-sync-tblgen`](https://github.com/Rot127/llvm-capstone/tree/auto-sync-tblgen?rgh-link-date=2023-03-10T22%3A14%3A54Z), except for `TriCoreGenDisassemblerTables.inc`, which modified a small part of `DecodeInstruction` and `DecodeToMCInst`

XVilka commented:
Looking forward for this to be merged.

Rot127 commented:
@imbillow Please open an issue after this is merged that TriCore should use the more general functions to copy the mapping data for instructions.
Just added the functions with this commit: https://github.com/capstone-engine/capstone/pull/1949/commits/518d71916b35f670050bc5f44a4fd9e58ceec3be

b1llow committed:
fix tricore endian

XVilka commented:
@kabeor could this be merged?

kabeor commented:
@XVilka need one more day to see if @aquynh has more thoughts.

XVilka commented:
I suppose no additional thoughts 😄

Rot127 commented:
@imbillow Please check if TriCore is affected by https://github.com/Rot127/llvm-capstone/commit/4955bf005bba0ed1979446b60373e6d59416eda6
(I doubt it but better check this now than by an issue weeks later).

b1llow committed:
fix: support for TriCore call group mapping

b1llow commented:
> @imbillow Please check if TriCore is affected by [Rot127/llvm-capstone@4955bf0](https://github.com/Rot127/llvm-capstone/commit/4955bf005bba0ed1979446b60373e6d59416eda6) (I doubt it but better check this now than by an issue weeks later).

It seems to have little effect, for the tricore side.
Because the generated disassembler seems to use only the `DecoderTable` and `checkDecoderPredicate()` in the `TriCoreGenDisassemblerTables.inc` file to determine the `subtarget`

b1llow committed:
fix all tricore compile warnings

XVilka commented:
@kabeor @aquynh sorry for being an annoyance, but it would be nice to be merged sooner than later, so the further work could be continued.

kabeor commented:
some thoughts from @aquynh:

> we need to document the llvm version, so later it is easier to track
> 
> i prefer to follow what other archs are doing, that is without adding inc/ directory
>
> the code look ok, but i will need to look at all once again on anther machine where i can comment

He will comment more later :)

XVilka commented:
As we noted before, there's no LLVM for Tricore, these tablegen files are "free-standing". And the autogeneration scripts that were used are part of auto-sync that is being worked separately in other PRs.

Rot127 commented:
> He will comment more later :)

Looking forward to it :) But here already some notes.

> i prefer to follow what other archs are doing, that is without adding inc/ directory

Other auto-sync archs will use an additional `inc` directory (in `include/capstone/`). This is necessary because we have to (or very much should) generated certain enums. But putting these enums in the `arch/<ARCH>` dir will break the architecture. Because core code (`cs.c` and the like) then needs to include headers from `arch/<ARCH>/`. Which is not wise IMHO.

> we need to document the llvm version, so later it is easier to track

This is in the making and will be added in the next week. But as XVilka said it is not really informative in case of TriCore.

aquynh commented:
On Thu, Apr 20, 2023, 18:13 Anton Kochkov ***@***.***> wrote:

> As we noted before, there's no LLVM for Tricore, these tablegen files are
> "free-standing". And the autogeneration scripts that were used are part of
> auto-sync that is being worked separately in other PRs.
>

We can host a copy of your llvm in our github organization in that case.

>
> Message ID: ***@***.***>
>

XVilka commented:
This LLVM doesn't exist, these tablegen files were written independently and separately by @imbillow

b1llow commented:
Yes, I only wrote the tablegen definition for tricore ISA and did not write the LLVM part, because the main purpose was to add tricore support to Capstone.

aquynh commented:
please put all the header code under include/capstone/inc/ into one file, just like the other archs.

aquynh commented on arch/TriCore/TriCoreDisassembler.c:
```diff
@@ -0,0 +1,1474 @@
+#include "../../MCDisassembler.h"
+#include "../../MathExtras.h"
+
+static bool readInstruction16(const uint8_t *code, size_t code_len, uint16_t *insn) {
```
please follow the coding style of existing code: that is to have the open bracket { of the functions on the next line, but at the end of the line like this.

aquynh commented on bindings/python/capstone/__init__.py:
```diff
@@ -174,7 +182,9 @@
 CS_ARCH_BPF = 14
 CS_ARCH_RISCV = 15
-CS_ARCH_MAX = 16
+# CS_ARCH_SH = 16
```
what happens to the SH arch here?

b1llow replied:
Because `CS_ARCH_TRICORE=17` corresponds to C code. Then there is a missing `CS_ARCH_SH=16` in between, but actually `CS_ARCH_SH` does not have Python bindings.

aquynh replied:
> > please put all the header code under include/capstone/inc/ into one file, just like the other archs.
> 
> @aquynh Could you point out what the advantage of this is? It kinda breaks the idea of `auto-sync` that generated content and hand written content is clearly separated into unique files.

i prefer to have it consistent with other archs.

is it too hard to update your tool to put all of the inc files into one?

Rot127 replied:
Its not really hard to change that. But it goes against the `auto-sync` architecture and is more a question of principle.

The update process is almost completely automated but internally complicated. So IMHO it is important that each update step of a file, is as isolated as much as possible from other files.

This is important because different files need different update mechanisms:

| Method | File type | Update method | Edits by hand |
|-------|------|----------------------|------------------------|
| 1 | `.inc` files | Exclusively generated by patched LLVM backends | Never/Not allowed |
| 2 | Translated LLVM C++ files | Translated by `CppTranslater` and diffing to previous file. | Rarely. Only changes which are too complicated for automation or diffing |
|3 |CS exclusive files | By hand | all |

The less dependencies these files have during the update process, the simpler `auto-sync` is.
Patching generated content (Method 1) into a by hand edited file (Method 3), adds such an dependency and more code.

> i prefer to have it consistent with other archs.

I agree that this is important, but since all LLVM based architectures (with the exception of `x86`) will be updated to use `auto-sync`, they are consistent to each other. This will add up to: `10/18 auto-sync archs` (TriCore included). And in the future even more, since `auto-sync` also eases the adding of new archs (as this PR already proved).

kabeor replied:
> since all LLVM based architectures (with the exception of x86) will be updated to use auto-sync, they are consistent to each other.

I agree with this.

b1llow replied:
I have an idea, maybe we can do it like this:

we insert comments like this in `<arch>.h`:

```
// generate content <section> begin
...
// generate content <section> end
```

tablgen replaces these comments with the actual content when running.

Rot127 replied:
Something like this would be the way I'd implement this.
But implementing is not so much a problem.
It is the decision what is more important.

- A single header file for each architecture.

Or

- Following the auto-sync rules for file updating.

XVilka replied:
I personally think we should switch to the new scheme, especially since there are plans to update other architectures, and add new ones (e.g. Xtensa, LANAI). The more architectures capstone has, the harder it will become to combine autogenerated and manual content in the same file.

XVilka replied:
@aquynh so what is your decision?

kabeor replied:
@XVilka Well, he insist:)

aquynh commented on cstool/cstool_tricore.c:
```diff
@@ -0,0 +1,65 @@
+
+void print_insn_detail_tricore(csh handle, cs_insn *ins)
+{
+  cs_tricore *tricore;
```
please use TABs for indentation, like existing code.

aquynh commented:
please fix the coding style, then we can be ready to merge.

Rot127 commented:
> please put all the header code under include/capstone/inc/ into one file, just like the other archs.

@aquynh Could you point out what the advantage of this is? It kinda breaks the idea of `auto-sync` that generated content and hand written content is clearly separated into unique files.

b1llow committed:
Format all .(c|h) code

b1llow commented:
Complete code formatting with clang-format (15.0.7) and https://github.com/torvalds/linux/blob/master/.clang-format

kabeor commented:
@aquynh

aquynh commented on arch/TriCore/TriCoreInstPrinter.c:
```diff
@@ -54,106 +55,113 @@ static inline void fill_mem(cs_tricore *tc, uint8_t base, int32_t disp);
+		case TriCore_LHA_abs: {
+			return false;
+		}
 		}
```
the { above does not align properly?

aquynh commented on arch/TriCore/TriCoreInstPrinter.c:
```diff
@@ -327,42 +349,43 @@ static void printDisp24Imm(MCInst *MI, int OpNum, SStream *O) {
+		default:
+			// handle other cases, if any
+			break;
 		}
```
this close bracket { does not align?

aquynh replied:
ok, it seems you align case under the same column with switch.

nvm then

aquynh commented:
code indentation looks better now :-)

Rot127 commented:
This would be the way I'd implement this.
But implementing is not so much a problem.
It is the decision what is more important.

- A single header file for each architecture.

Or

- Following the auto-sync rules for file updating.

22 abr 2023 6:54:59 billow ***@***.***>:

> 
> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In bindings/python/capstone/__init__.py[https://github.com/capstone-engine/capstone/pull/1973#discussion_r1174290822]:
> 
>> @@ -174,7 +182,9 @@
>  CS_ARCH_WASM = 13
>  CS_ARCH_BPF = 14
>  CS_ARCH_RISCV = 15
> -CS_ARCH_MAX = 16
> +# CS_ARCH_SH = 16
> 
> I have an idea, maybe we can do it like this:
> 
> we insert comments like this in *<arch>.h*:
> 
> *// generate content <section> begin
> ...
> // generate content <section> end
> *
> tablgen replaces these comments with the actual content when running.
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/capstone-engine/capstone/pull/1973#discussion_r1174290822], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AK5ET6AVPCJQECOJC5HSBFLXCNQCDANCNFSM6AAAAAAVXABL74].
> You are receiving this because you were mentioned.[Imagen de rastreo][https://github.com/notifications/beacon/AK5ET6B7AQXPBFWQBINNCV3XCNQCDA5CNFSM6AAAAAAVXABL76WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSTHXWCU.gif]
>

b1llow commented:
So I need to manually **insert** the contents of the `inc` files into the `.h` file in this PR?

Rot127 commented:
> So I need to manually insert the contents of the inc files into the .h file in this PR?

If @aquynh says so, yes. Hadn't the time to write a patch script yet.

b1llow committed:
Fix format error

b1llow committed:
Fix `print_insn_detail_tricore` and disp fill

XVilka reviewed (changes requested):
@imbillow could you please also update `HACK.TXT`?

b1llow committed:
Fix tricore python binding

- fix HACK.TXT

XVilka commented on HACK.TXT:
```diff
@@ -14,11 +14,13 @@ Capstone source is organized as followings.
-│   ├── TMS320C64x  <- TMS320C64x engine
-│   ├── X86         <- X86 engine
-│   └── XCore       <- XCore engine
+│   ├── RISCV <- RISCV engine
```
Seems like broken indent

b1llow committed:
Fix HACK.TXT

b1llow committed:
Fix tricore.h and remove `inc` folder

XVilka commented:
One more argument against mixing autogenerated parts and manual parts inside one file - the risk of someone modifying the autogenerated part without noticing it, and `git` wouldn't be able to differentiate (e.g. [through `.gitattributes`](https://medium.com/@clarkbw/managing-generated-files-in-github-1f1989c09dfd)) between these. Separating them could help preventing this.

XVilka commented on include/capstone/tricore.h:
```diff
@@ -0,0 +1,582 @@
+	// generate content <TriCoreGenCSFeatureEnum.inc> begin
+	// clang-format off
+
+/* Capstone Disassembly Engine, https://www.capstone-engine.org */
```
Should we also remove the header here? I also wonder if maybe it's better to also follow the indent here

XVilka replied:
@kabeor @aquynh @Rot127  do you agree with me? I think it's the only thing left before it could be merged, right?

Rot127 commented:
@XVilka Ah yes. Of cause. Removing the header is better.

26 abr 2023 4:47:29 Anton Kochkov ***@***.***>:

> 
> ***@***.**** commented on this pull request.
> 
> ----------------------------------------
> 
> In include/capstone/tricore.h[https://github.com/capstone-engine/capstone/pull/1973#discussion_r1177280435]:
> 
>> +//> Group of TriCore instructions
> +typedef enum tricore_insn_group {
> +   TriCore_GRP_INVALID, ///< = CS_GRP_INVALID
> +   //> Generic groups
> +   TriCore_GRP_CALL, ///< = CS_GRP_CALL
> +   TriCore_GRP_JUMP, ///< = CS_GRP_JUMP
> +   TriCore_GRP_ENDING, ///< = mark the end of the list of groups
> +} tricore_insn_group;
> +
> +typedef enum tricore_feature_t {
> +
> +   TriCore_FEATURE_INVALID = 0,
> +   // generate content <TriCoreGenCSFeatureEnum.inc> begin
> +   // clang-format off
> +
> +/* Capstone Disassembly Engine, https://www.capstone-engine.org */
> 
> @kabeor[https://github.com/kabeor] @aquynh[https://github.com/aquynh] @Rot127[https://github.com/Rot127] do you agree with me? I think it's the only thing left before it could be merged, right?
> 
> —
> Reply to this email directly, view it on GitHub[https://github.com/capstone-engine/capstone/pull/1973#discussion_r1177280435], or unsubscribe[https://github.com/notifications/unsubscribe-auth/AK5ET6AZWD3UOT4MA2YE6X3XDCED7ANCNFSM6AAAAAAVXABL74].
> You are receiving this because you were mentioned.[Imagen de rastreo][https://github.com/notifications/beacon/AK5ET6CU5SACLYEHRSDNQUTXDCED7A5CNFSM6AAAAAAVXABL76WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTSTQKUUG.gif]
>

b1llow committed:
Fix tricore.h

XVilka reviewed (approved):
LGTM. @aquynh @kabeor all your feedback was addressed. Your call now.

kabeor commented on include/capstone/tricore.h:
```diff
@@ -0,0 +1,562 @@
+	// generate content <TriCoreGenCSRegEnum.inc> begin
+	// clang-format off
+
+	TriCore_REG_INVALID = 0,
```
please capitalize:)

b1llow replied:
@Rot127 Maybe we need to capitalize the target name when generating `Enum` in Capstone exports?

Rot127 replied:
Yes. Its on the list.

Rot127 replied:
@imbillow I pushed a commit which should capitalize the target name for all enums. Can you please test it and create a PR if something doesn't work.

b1llow replied:
There is nothing wrong with most of them, except for these two places. I will come to PR later if I have time.

- [x] `TriCoreGenDisassemblerTables.inc#checkDecoderPredicate`
- [x] `TriCoreGenCSRegEnum.inc`
- [x] `TriCoreGenRegisterInfo.inc`

kabeor commented:
Then we can merge

kabeor commented on bindings/python/capstone/tricore.py:
```diff
@@ -0,0 +1,45 @@
+import ctypes, copy
+from .tricore_const import *
+
+class TriCoreOpMem(ctypes.Structure):
```
and plz capital here

b1llow replied:
> and plz capital here

![image](https://user-images.githubusercontent.com/24983369/234862695-de06fdb4-d109-4b45-a29d-65e5ca0217d0.png)

But I noticed that PowerPC isn't capitalized either. I think there might not be such a convention?

Rot127 replied:
Thinking about it a second time I'm against capitalizing it. We would need to add `.upper()` pretty much everywhere in the LLVM backends when using the target name (which is done a lot). So the code gets more unreadable.
@kabeor Is there any advantage for capitalization which would make this worth it?

kabeor replied:
Oh sorry, my bad. Here is good.

kabeor commented on bindings/python/capstone/tricore_const.py:
```diff
@@ -0,0 +1,473 @@
+TRICORE_OP_IMM = 2
+TRICORE_OP_MEM = 3
+
+TriCore_REG_INVALID = 0
```
here

b1llow committed:
Merge branch 'next' into tricore

Rot127 commented on suite/test_corpus3.py:
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/python
```
```suggestion
#!/usr/bin/env python3
```

b1llow committed:
Upper all `inc` and fix

b1llow committed:
Upper tricore_const.py

b1llow committed:
Update suite/test_corpus3.py

Co-authored-by: Rot127 <45763064+Rot127@users.noreply.github.com>

b1llow commented:
@kabeor Hi, the problem you mentioned has been fixed, please review again.

Rot127 commented:
@kabeor @aquynh Is it ready to be merged?

kabeor commented:
@imbillow @Rot127 It's really great, can you solve the conflict so I can merge?

b1llow committed:
Merge branch 'next' into tricore

XVilka commented:
@kabeor looks like conflict was resolved 😁

kabeor commented:
Thanks again for this great work! Merged.

XVilka commented:
@aquynh @kabeor what do you think about tagging 5.0-rc3 with that (and release 5.0 once it's stable enough)? I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?
cc @imbillow @Rot127 

Maybe addressing these two first though: https://github.com/capstone-engine/capstone/issues/2010 and https://github.com/capstone-engine/capstone/issues/2009

aquynh commented:
this is what i am planning to do.

anything else do you think should be merged before rc3 tag?

aquynh commented:
> @aquynh @kabeor what do you think about tagging 5.0-rc3 with that (and release 5.0 once it's stable enough)? I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work? cc @imbillow @Rot127
> 
> Maybe addressing these two first though: #2010 and #2009

yes, please make pull req for the above issues before rc3 tag.

my main concern now is that the Pypi package is still properly generated for all platforms.

Rot127 commented:
> I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?

@aquynh Would agree with releasing `v5.0` before `auto-sync` and `v5.1` for updated `ARM`, `PPC` and whatever else is done at this point.

aquynh commented:
> > I wonder if it make sense to release 5.0 before auto-sync is merged, and focus 5.1 on auto-sync work?
> 
> @aquynh Would agree with releasing `v5.0` before `auto-sync` and `v5.1` for updated `ARM`, `PPC` and whatever else is done at this point.

yes the auto-sync is a big change, so I want to release 5.0 before merging that.

peace-maker commented on bindings/python/capstone/__init__.py:
```diff
@@ -708,6 +726,8 @@ def __gen_detail(self):
         elif arch == CS_ARCH_RISCV:
             (self.operands) = riscv.get_arch_info(self._raw.detail.contents.arch.riscv)
+        elif arch == CS_ARCH_TRICORE:
+            (self.operands) = riscv.get_arch_info(self._raw.detail.contents.arch.tricore)
```
Should this be `tricore.get_arch_info` instead of `riscv.get_arch_info`?

b1llow replied:
You are right

b1llow replied:
@peace-maker Could you please send a new PR to fix these errors?
